### PR TITLE
Add assert to Unix GCToOSInterface::GetCurrentProcessorNumber for clarity

### LINKED ISSUE #31995: Thread.GetCurrentProcessorId may return wrong value if HAVE_SCHED_GETCPU is not set
If `HAVE_SCHED_GETCPU` is not set, then `GetCurrentProcessorNumber` will return 0.

https://github.com/dotnet/runtime/blob/fcd862e06413a000f9cafa9d2f359226c60b9b42/src/coreclr/src/gc/unix/gcenv.unix.cpp#L481

However, the code here assume that we'll get -1, which we won't. See:

https://github.com/dotnet/runtime/blob/aedf8f52006619ef5d4eca65d79f42cc4b7bc402/src/coreclr/src/System.Private.CoreLib/src/System/Threading/ProcessorIdCache.cs#L35

### PR DESCRIPTION
**Pull request by jkotas**
Fixes #31995

### CODE REVIEW

File: src/coreclr/src/gc/unix/gcenv.unix.cpp [hunk 1]
```diff
@@ -478,6 +478,7 @@ uint32_t GCToOSInterface::GetCurrentProcessorNumber()
     assert(processorNumber != -1);
     return processorNumber;
 #else
+    assert(false); // This method is expected to be called only if CanGetCurrentProcessorNumber is true
```
Comment (lpereira) [inline src/coreclr/src/gc/unix/gcenv.unix.cpp]: Maybe `assert(!CanGetCurrentProcessorNumber())` instead and get rid of the comment?

**Review by janvorli — APPROVED** [review]

File: src/coreclr/src/gc/unix/gcenv.unix.cpp (hunk 1 above)
Comment (janvorli) [reply src/coreclr/src/gc/unix/gcenv.unix.cpp]: I would then rather add the same assert (`assert(CanGetCurrentProcessorNumber());`) as we have in the other versions of this function at the beginning of the function.

Comment (jkotas) [reply src/coreclr/src/gc/unix/gcenv.unix.cpp]: I had the assert at the beginning of the function originally. It did not seem to make the intent obvious. That's why I have added the comment instead.

Comment (janvorli) [reply src/coreclr/src/gc/unix/gcenv.unix.cpp]: I am fine with both ways.

**Review by stephentoub — APPROVED** [review]

**Pull request merged**